modules/images/webp-uploads/generation.php

```diff
@@ -0,0 +1,236 @@
+			// If the current size exactly matches requested width and height.
+			( $width === $img_width && $height === $img_height ) ||
+			// If the current size matches the requested height and has a variable width.
+			( 0 === $width && $height === $img_height ) ||
+			// If the current size matches the requested width and has a variable height.
+			( $width === $img_width && 0 === $height )
+		) {
+			$size = $size_name;
+
+			// Stop finding the size name if the current size exactly matches requested width and height.
+			// If the current size has a variable width or height, we need to continue finding in case there
+			// is an exact match.
+			if ( $width && $height ) {
+				break;
+			}
+		}
+	}
+
+	// Bail if the subsize name is not found.
+	if ( empty( $size ) ) {
+		return;
+	}
+
+	// Bail if we can't find an attachment with requested file.
+	$attachment_id = attachment_url_to_postid( "{$uploads_dir['baseurl']}/{$matches[1]}.{$matches[4]}" );
+	if ( empty( $attachment_id ) ) {
+		return;
+	}
+
+	// Bail if we can't find attachment metadata.
+	$metadata = wp_get_attachment_metadata( $attachment_id );
+	if ( empty( $metadata ) ) {
+		return;
+	}
+
+	require_once ABSPATH . '/wp-admin/includes/image.php';
+
+	// Generate missing sub size.
+	$metadata = _wp_make_subsizes(
+		array( $size => $sizes[ $size ] ),
+		"{$uploads_dir['basedir']}/{$matches[1]}.{$matches[4]}",
+		$metadata,
+		$attachment_id
+	);
+
+	// Generate sources.
+	$metadata = webp_uploads_create_sources_property( $metadata, $attachment_id );
+
+	// Save updated metadata.
+	wp_update_attachment_metadata( $attachment_id, $metadata );
+
+	$image_file = $uploads_dir['basedir'] . $relpath;
+
+	// Send headers if headers haven't been sent yet.
+	if ( ! headers_sent() ) {
+		$filetype = wp_check_filetype( $image_file );
+		if ( ! empty( $filetype['type'] ) ) {
+			header( 'Content-type: ' . $filetype['type'] );
+		}
+
+		header( 'Content-Length: ' . filesize( $image_file ) );
+		header( 'Cache-Control: public, max-age=' . 0 );
+		header( 'Content-Disposition: filename=' . wp_basename( $image_file ) );
+		header( 'Last-Modified: ' . gmdate( DATE_RFC2822 ) );
+	}
+
+	// Send newly generated file and exit.
+	readfile( $image_file );
+	exit;
+}
+add_action( 'template_redirect', 'webp_uploads_generate_missing_images' );
+
+/**
+ * Generates srcset URLs for missing sizes.
+ *
+ * @since n.e.x.t
+ *
+ * @param array  $sources       One or more arrays of source data to include in the 'srcset'.
+ * @param array  $size_array    An array of requested width and height values.
+ * @param string $image_src     The 'src' of the image.
+ * @param array  $image_meta    The image meta data as returned by 'wp_get_attachment_metadata()'.
+ * @param int    $attachment_id Image attachment ID or 0.
+ * @return array Srcset sources array.
+ */
+function webp_uploads_generate_image_srcset( $sources, $size_array, $image_src, $image_meta, $attachment_id ) {
+	$image_src_parts = wp_parse_url( $image_src );
```

Comment:
We should make sure the returned value from `wp_parse_url` is not `false` or make sure is a string.